packer/builder/azure/arm/builder.go

```diff
@@ -0,0 +1,140 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See the LICENSE file in the project root for license information.
+
+package arm
+
+import (
+	"log"
+
+	"github.com/Azure/packer-azure/packer/builder/azure/common/constants"
+	"github.com/Azure/packer-azure/packer/builder/azure/common/lin"
+
+	"github.com/Azure/azure-sdk-for-go/Godeps/_workspace/src/github.com/Azure/go-autorest/autorest/azure"
+
+	"github.com/mitchellh/multistep"
+	"github.com/mitchellh/packer/common"
+	"github.com/mitchellh/packer/helper/communicator"
+	"github.com/mitchellh/packer/packer"
+)
+
+type Builder struct {
+	config   *Config
+	stateBag multistep.StateBag
+	runner   multistep.Runner
+}
+
+const (
+	DefaultPublicIPAddressName = "sshPublicIP"
+)
+
+func (b *Builder) Prepare(raws ...interface{}) ([]string, error) {
+	c, warnings, errs := newConfig(raws...)
+	if errs != nil {
+		return warnings, errs
+	}
+
+	b.config = c
+
+	b.stateBag = new(multistep.BasicStateBag)
+	err := b.configureStateBag(b.stateBag)
+	if err != nil {
+		return nil, err
+	}
+
+	return warnings, errs
+}
+
+func (b *Builder) Run(ui packer.Ui, hook packer.Hook, cache packer.Cache) (packer.Artifact, error) {
+	ui.Say("Preparing builder ...")
+	ui.Message("Creating Azure Resource Manager (ARM) client ...")
+
+	// TODO(chrboum): figure out how to implement a test case to validate these properties
+	b.stateBag.Put("hook", hook)
+	b.stateBag.Put(constants.Ui, ui)
+
+	servicePrincipalToken, err := b.createServicePrincipalToken()
+	if err != nil {
+		return nil, err
+	}
+
+	azureClient, err := NewAzureClient(b.config.SubscriptionID, b.config.ResourceGroupName, b.config.StorageAccount, servicePrincipalToken)
+	if err != nil {
+		return nil, err
+	}
+
+	steps := []multistep.Step{
+		NewStepCreateResourceGroup(azureClient, ui),
+		NewStepValidateTemplate(azureClient, ui),
+		NewStepDeployTemplate(azureClient, ui),
+		NewStepGetIPAddress(azureClient, ui),
+		&communicator.StepConnectSSH{
+			Config:    &b.config.Comm,
+			Host:      lin.SSHHost,
+			SSHConfig: lin.SSHConfig(b.config.UserName),
+		},
+		&common.StepProvision{},
+		&lin.StepGeneralizeOS{
+			Command: "sudo /usr/sbin/waagent -force -deprovision+user && export HISTSIZE=0 && sync",
+		},
+		NewStepQueryVM(azureClient, ui),
```

Comment:
yeah, as a reader, I'm not sure what we're doing here... (querying... what or why?)